### PR TITLE
Add warnings about NFS

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

**Comment by roidelapluie** [comment]
Also see related https://github.com/prometheus/docs/pull/1706

File: docs/storage.md [hunk 1]
```diff
@@ -73,7 +73,9 @@ needed_disk_space = retention_time_seconds * ingested_samples_per_second * bytes
...
+If your local storage becomes corrupted for whatever reason, your best bet is to shut down Prometheus and remove the entire storage directory. You can try removing individual block directories, or WAL directory, to resolve the problem, this means losing a time window of around two hours worth of data per block directory. Again, Prometheus's local storage is not meant as durable long-term storage.
+
+CAUTION: Non POSIX compliant filesystems are not supported by Prometheus's local storage, corruptions may happen, without possibility to recover. NFS is only potentially POSIX, most implementations are not. Cases of corruption are also reported with other networked filesystems. It is strongly recommended to use a local filesystem for reliability.
```
Comment (RichiH) [inline docs/storage.md:78]:
```suggestion
CAUTION: Non-POSIX compliant filesystems are not supported by Prometheus' local storage as unrecoverable corruptions may happen. NFS could be POSIX-compliant, but most implementations are not. Cases of corruption are also reported with other networked filesystems. It is strongly recommended to use a local filesystem for reliability.
```

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -696,7 +696,11 @@ func main() {
 					return errors.Wrapf(err, "opening storage failed")
 				}
 
-				level.Info(logger).Log("fs_type", prom_runtime.Statfs(cfg.localStoragePath))
+				if fsType := prom_runtime.Statfs(cfg.localStoragePath); fsType == "NFS_SUPER_MAGIC" {
+					level.Warn(logger).Log("fs_type", fsType, "msg", "This filesystem is not supported and may lead to data corruption and data loss. Please read carefully https://prometheus.io/storage to learn more about supported filesystems.")
```
Comment (brian-brazil) [inline cmd/prometheus/main.go:700]: carefully read

File: docs/storage.md [hunk 3]
```diff
@@ -75,7 +75,7 @@ To tune the rate of ingested samples per second, you can either reduce the numbe
 
 If your local storage becomes corrupted for whatever reason, your best bet is to shut down Prometheus and remove the entire storage directory. You can try removing individual block directories, or WAL directory, to resolve the problem, this means losing a time window of around two hours worth of data per block directory. Again, Prometheus's local storage is not meant as durable long-term storage.
 
-CAUTION: Non-POSIX compliant filesystems are not supported by Prometheus' local storage as unrecoverable corruptions may happen. NFS is could be POSIX-compliant, but most implementations are not. Cases of corruption are also reported with other networked filesystems. It is strongly recommended to use a local filesystem for reliability.
+CAUTION: Non-POSIX compliant filesystems are not supported by Prometheus' local storage as unrecoverable corruptions may happen. NFS could be POSIX-compliant, but most implementations are not. Cases of corruption are also reported with other networked filesystems. It is strongly recommended to use a local filesystem for reliability.
```
Comment (brian-brazil) [inline docs/storage.md:78]: Should we explicitly also call out AWS EFS, as many users of it don't seem to realise it's NFS?

Comment (roidelapluie) [reply docs/storage.md]: Yes, noted.

File: docs/storage.md [hunk 4]
```diff
@@ -73,7 +73,9 @@ needed_disk_space = retention_time_seconds * ingested_samples_per_second * bytes
...
+If your local storage becomes corrupted for whatever reason, your best bet is to shut down Prometheus and remove the entire storage directory. You can try removing individual block directories, or WAL directory, to resolve the problem, this means losing a time window of around two hours worth of data per block directory. Again, Prometheus's local storage is not meant as durable long-term storage.
+
+CAUTION: Non-POSIX compliant filesystems are not supported by Prometheus' local storage as unrecoverable corruptions may happen. NFS could be POSIX-compliant, but most implementations are not. Notably, AWS EFS is known to cause corruptions. Cases of corruption are also reported with other networked filesystems. It is strongly recommended to use a local filesystem for reliability.
```
Comment (brian-brazil) [inline docs/storage.md:78]: That's a bit harsh on EFS - we don't have definitive proof but there seems to be a fair few users using it who have had corruption. I'd weaken to something like "NFS filesystems (including AWS's EFS) are not supported".

> Cases of corruption are also reported with other networked filesystems.

It's not yet clear if SMB is an issue, so I'd leave this out for now.

Comment (roidelapluie) [reply docs/storage.md:78]: We do have a report today about glusterfs.

Comment (roidelapluie) [reply docs/storage.md:78]: (on the mailing list I think)

Comment (roidelapluie) [reply docs/storage.md]: I have removed the sentence. We can still adapt later.

**Comment by roidelapluie** [comment]
I have tested an EC2 instance with EFS and even when you `mount -t efs` it is using nfs, so we see NFS_SUPER_MAGIC fs_type.

**Comment by brian-brazil** [comment]
:+1:

**Comment by roidelapluie** [comment]
@brian-brazil I have moved to switch{} for better readability in main.go

**Comment by brian-brazil** [comment]
:+1:

**Pull request merged**